### Data Source
**Dataset:** Online Retail II \
**Source:** [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/502/online+retail+ii)

### Dataset Description
This dataset contains all the transactions occurring for a UK-based and registered, non-store online retailer between 01/12/2009 and 09/12/2011. The company mainly sells unique all-occasion gift-ware, and many of its customers are wholesalers. The data spans two full years and is provided as two Excel sheets, one per year.

The dataset contains 1,067,371 instances across 8 features, characterized as multivariate, sequential, time-series data with both integer and real-valued fields. It is associated with three types of machine learning tasks: Classification, Regression, and Clustering, which is directly relevant to this project's planned modeling approach (CLV regression, churn classification, customer segmentation via clustering).

**Columns:**
- `Invoice` — 6-digit invoice number; a prefix of 'C' indicates a cancelled transaction
- `StockCode` — 5-digit product code
- `Description` — product name
- `Quantity` — units purchased per transaction
- `InvoiceDate` — date and time of transaction
- `Price` — unit price in GBP (£)
- `Customer ID` — 5-digit customer identifier
- `Country` — customer's country of residence


In [1]:
import pandas as pd
import sys
sys.path.append("../src")
from utils import report_reduction
from cleaning import net_cancellations

# List sheet names in the source Excel file to confirm its structure before loading
pd.ExcelFile("../data/online_retail_II.xlsx").sheet_names

['Year 2009-2010', 'Year 2010-2011']

In [2]:
# Load each yearly sheet separately
df_2009_2010 = pd.read_excel("../data/online_retail_II.xlsx", sheet_name="Year 2009-2010")
df_2010_2011 = pd.read_excel("../data/online_retail_II.xlsx", sheet_name="Year 2010-2011")

# Combine both years into a single dataframe covering the full date range
df = pd.concat([df_2009_2010, df_2010_2011], ignore_index=True)

# Confirm the combined row count matches UCI's stated total (1,067,371 instances)
df.shape

(1067371, 8)

### Check Data Types and Missing Values

With the full dataset loaded, we inspect column data types and non-null counts. This tells us which columns have missing values (and how many) before deciding how to handle them, and confirms whether any columns need dtype conversion.

In [3]:
# Check dtypes and non-null counts per column to identify missing values and confirm data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 78.8+ MB


### Quantify Missing Values

`df.info()` shows non-null counts, but it's clearer to explicitly compute the count and percentage of missing values per column. This confirms `Description` and `Customer ID` are the only columns with missing data, and shows the scale of each — which will inform how we handle them.

In [4]:
# Compute missing value count and percentage per column
missing_summary = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing %": (df.isnull().sum() / len(df) * 100).round(2)
})

# Only show columns that actually have missing values
missing_summary[missing_summary["Missing Count"] > 0]

,Missing Count,Missing %
Description,4382,0.41
Customer ID,243007,22.77


### Check for Duplicate Rows

Before investigating specific data quality issues like cancellations or invalid values, we check for fully duplicated rows. Duplicate transaction records — whether from data entry errors or export artifacts — could distort Frequency and Monetary calculations later if left unaddressed.

In [5]:
# Count fully duplicated rows across all columns
df.duplicated().sum()

np.int64(34335)

Duplicate rows account for a meaningful portion of the dataset (~3.2%). Before deciding whether to remove them, we inspect a sample to understand whether these represent genuine repeated line items (e.g., the same product added twice to an order) or artifacts from data export/loading.

In [6]:
# Inspect a sample of duplicated rows to understand their nature before deciding how to handle them
df[df.duplicated(keep=False)].sort_values(by=["Invoice", "StockCode"]).head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [7]:
# Remove exact duplicate rows, keeping the first occurrence
df = df.drop_duplicates().reset_index(drop=True)

In [8]:
# Confirm the new shape after removing duplicates
df.shape

(1033036, 8)

### Investigate Cancellations

Per the dataset documentation, an `Invoice` value prefixed with 'C' indicates a cancelled transaction. We quantify how many such rows exist before deciding how to handle them.

In [9]:
# Count rows flagged as cancellations (Invoice starts with 'C')
df['Invoice'].astype(str).str.startswith('C').sum()

np.int64(19104)

### Compare Cancellations to Negative Quantity Rows

Cancellations are identified by the 'C' prefix in `Invoice`, but negative `Quantity` values could also indicate returns, adjustments, or other non-purchase entries that aren't captured by this prefix alone. We compare the two to check for overlap and identify any additional rows that need investigation.

In [10]:
# Count rows with negative Quantity
negative_qty = (df['Quantity'] < 0).sum()

# Count rows flagged as cancellations
cancellations = df['Invoice'].astype(str).str.startswith('C').sum()

# Count rows with negative Quantity that are NOT flagged as cancellations
negative_qty_not_cancelled = ((df['Quantity'] < 0) & (~df['Invoice'].astype(str).str.startswith('C'))).sum()

print("Negative Quantity rows:", negative_qty)
print("Cancellation rows (Invoice starts with C):", cancellations)
print("Negative Quantity but NOT flagged as cancellation:", negative_qty_not_cancelled)

Negative Quantity rows: 22496
Cancellation rows (Invoice starts with C): 19104
Negative Quantity but NOT flagged as cancellation: 3393


As before, we inspect the `StockCode` values for rows with negative `Quantity` that are not flagged as cancellations, to understand whether these represent a distinct pattern (e.g., manual adjustment codes) or scattered data inconsistencies.

In [11]:
# Inspect StockCode values for negative-Quantity rows not flagged as cancellations
df[(df['Quantity'] < 0) & (~df['Invoice'].astype(str).str.startswith('C'))]['StockCode'].value_counts().head(15)

StockCode
22423     11
82494L     6
22719      6
46000M     5
84016      5
85017A     5
47566B     5
20852      5
85175      5
21830      5
71477      4
46000S     4
21768      4
84559D     4
84990      4
Name: count, dtype: int64

These rows carry real product codes but no 'C' prefix. Before deciding how to handle them, we check whether they are attributed to customers at all: if they have no `Customer ID`, they cannot affect customer-level features regardless of what they represent.

In [12]:
# Negative-Quantity rows without a 'C' invoice: are they attributed to customers, and at what price?
neg_not_c = df[(df['Quantity'] < 0) & (~df['Invoice'].astype(str).str.startswith('C'))]

print("Rows:", len(neg_not_c))
print("With a Customer ID:", neg_not_c['Customer ID'].notna().sum())
print("With Price > 0:", (neg_not_c['Price'] > 0).sum())
neg_not_c['Description'].value_counts().head(10)

Rows: 3393
With a Customer ID: 0
With Price > 0: 0


Description
check                     121
damages                    83
?                          81
damaged                    78
missing                    27
sold as set on dotcom      20
Damaged                    17
smashed                     9
thrown away                 9
Unsaleable, destroyed.      9
Name: count, dtype: int64

None of these rows has a `Customer ID` or a positive `Price`. Where a description is present, it records a stock adjustment rather than a customer return ("damaged", "smashed", "thrown away", "missing", "check"). These are warehouse write-offs and stock-count corrections. They are removed by the missing `Customer ID` filter and never reach customer-level analysis.

### Cancellations Must Be Netted, Not Just Dropped

A cancellation line (invoice prefixed 'C') reverses an earlier purchase. Dropping rows with `Quantity <= 0` removes the cancellation line but keeps the purchase it reversed, so revenue that was never received stays in the data.

Two orders show how large the effect can be. Both were placed and cancelled within minutes, yet a `Quantity > 0` filter would keep the full purchase value:

In [13]:
# The two largest cancelled orders: each purchase and its cancellation, minutes apart
pairs = df[
    df['Invoice'].astype(str).isin(['541431', 'C541433', '581483', 'C581484']) &
    df['StockCode'].astype(str).isin(['23166', '23843'])
]
pairs[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'InvoiceDate', 'Customer ID']]

,Invoice,StockCode,Description,Quantity,Price,InvoiceDate,Customer ID
557400,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04,2011-01-18 10:01:00,12346.0
557405,C541433,23166,MEDIUM CERAMIC TOP STORAGE JAR,-74215,1.04,2011-01-18 10:17:00,12346.0
1031554,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08,2011-12-09 09:15:00,16446.0
1031555,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2.08,2011-12-09 09:27:00,16446.0


Customer 12346's order of 74,215 storage jars (£77,183.60) was cancelled 16 minutes later, and customer 16446's order of 80,995 "Paper Craft, Little Birdie" units (£168,469.60) was cancelled 12 minutes later. Under a `Quantity > 0` filter, both would count as genuine revenue. The second order falls in the modelling holdout window, where it would become customer 16446's entire 6-month target value and flip their churn label from churned to returned.

Across the dataset, cancellations of product lines by identified customers total about 4.2% of gross sales value. They fall into three cases:

- **Exact pair (76% of cancelled value):** same customer, product, quantity and price as an earlier purchase.
- **Partial or non-exact (18%):** the customer bought the product earlier, but quantities or prices differ.
- **Unmatched (5%):** no earlier purchase exists in the data, typically because the original order predates December 2009.

**Netting rule.** Each cancellation, taken in chronological order, removes quantity from the same customer's earlier purchases of the same product, most recent first (LIFO). Purchases reduced to zero are dropped; cancelled quantity with no earlier purchase to offset is discarded. This handles all three cases with one rule. Recency is used as the matching signal because reversals typically follow soon after the order; the case of customer 15098, shown after the cleaning steps below, illustrates why matching on price would be wrong.

The implementation is `net_cancellations()` in `src/cleaning.py`.

### Check for Invalid Price Values

Beyond Quantity issues, we check for rows with zero or negative `Price`, which would not represent genuine revenue-generating transactions.

In [14]:
# Count rows with zero or negative Price
(df['Price'] <= 0).sum()

np.int64(6019)

### Check for Non-Product StockCode Entries

Some rows have a `StockCode` that does not represent merchandise: administrative, adjustment or placeholder entries rather than product sales. Genuine product codes follow a consistent format of five digits, optionally followed by one or two letters (e.g. `21491`, `84951A`, `15056BL`). Codes outside this format were reviewed individually by their descriptions.

Excluded as non-merchandise: `POST` and `DOT` (postage), `M` (manual entries), `C2` (carriage), `D` (discounts), `BANK CHARGES`, `ADJUST` and `ADJUST2` (manual accounting adjustments), `TEST001` and `TEST002` (test products), and `PADS` (a £0.001 placeholder line). One non-standard code, `SP1002` (a kid's chalkboard/easel), is a genuine product and is kept.

The list is defined once, as `NON_PRODUCT_CODES` in `src/config.py`, so every notebook uses the same definition.

In [15]:
from config import NON_PRODUCT_CODES
non_product_codes = NON_PRODUCT_CODES

# Row count and an example description per non-product code, as evidence for each exclusion
flagged = df[df['StockCode'].astype(str).isin(non_product_codes)]
flagged.groupby(flagged['StockCode'].astype(str)).agg(
    Rows=('Invoice', 'size'),
    With_Customer_ID=('Customer ID', 'count'),
    Example_Description=('Description', 'first')
).sort_values('Rows', ascending=False)

,Rows,With_Customer_ID,Example_Description
StockCode,,,
POST,2086,1983,POSTAGE
DOT,1425,16,DOTCOM POSTAGE
M,1387,1085,Manual
C2,277,254,CARRIAGE
D,173,170,Discount
BANK CHARGES,100,37,Bank Charges
ADJUST,67,61,Adjustment by john on 26/01/2010 16
PADS,19,19,PADS TO MATCH ALL CUSHIONS
TEST001,15,15,This is a test product.


These codes account for a small share of rows, but several carry customer IDs and non-trivial values (manual adjustments of up to several hundred pounds each), so leaving them in would inflate specific customers' Monetary values. They are excluded in the cleaning steps below.

### Apply Cleaning Steps

Filters are applied one at a time so each step's impact is reported separately. Order matters:

1. **Drop missing `Customer ID`:** transactions that cannot be attributed to a customer are unusable for CLV.
2. **Drop non-product `StockCode` entries:** postage, fees and manual adjustments are not merchandise sales.
3. **Drop `Price <= 0`:** applied *before* netting, so a cancellation can never be offset against a free £0 line instead of the paid purchase it reversed.
4. **Net cancellations against purchases (LIFO):** removes all cancellation lines and reduces or removes the purchases they reversed. This replaces the earlier `Quantity > 0` filter, which removed cancellation lines but kept the reversed purchases.

In [16]:
start = df.shape

before = df.shape
df_clean = df[df['Customer ID'].notna()].copy()
report_reduction(before, df_clean.shape, step_name="Drop missing Customer ID")

before = df_clean.shape
df_clean = df_clean[~df_clean['StockCode'].isin(non_product_codes)]
report_reduction(before, df_clean.shape, step_name="Drop non-product StockCodes")

before = df_clean.shape
df_clean = df_clean[df_clean['Price'] > 0]
report_reduction(before, df_clean.shape, step_name="Drop Price <= 0")

before = df_clean.shape
df_clean, netting_summary = net_cancellations(df_clean)
report_reduction(before, df_clean.shape, step_name="Net cancellations (LIFO)")

print("Netting detail:")
for k, v in netting_summary.items():
    print(f"  {k:32s} {v:,}")

print()
report_reduction(start, df_clean.shape, step_name="All cleaning steps combined")

Drop missing Customer ID:
  Before: 1,033,036 rows
  After:  797,885 rows
  Removed: 235,151 rows (22.76%)
Drop non-product StockCodes:
  Before: 797,885 rows
  After:  794,241 rows
  Removed: 3,644 rows (0.46%)
Drop Price <= 0:
  Before: 794,241 rows
  After:  794,181 rows
  Removed: 60 rows (0.01%)
Net cancellations (LIFO):
  Before: 794,181 rows
  After:  770,554 rows
  Removed: 23,627 rows (2.98%)
Netting detail:
  cancellation_lines               17,602
  fully_netted                     15,877
  partly_netted                    126
  unmatched                        1,599
  value_netted                     730,406.29
  value_unmatched_discarded        42,065.74
  purchase_rows_reduced            15,708
  purchase_rows_removed_entirely   6,025

All cleaning steps combined:
  Before: 1,033,036 rows
  After:  770,554 rows
  Removed: 262,482 rows (25.41%)


In [17]:
# Confirm no missing values remain and check dtypes
df_clean.info()

<class 'pandas.DataFrame'>
Index: 770554 entries, 0 to 1033034
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      770554 non-null  object        
 1   StockCode    770554 non-null  object        
 2   Description  770554 non-null  object        
 3   Quantity     770554 non-null  int64         
 4   InvoiceDate  770554 non-null  datetime64[us]
 5   Price        770554 non-null  float64       
 6   Customer ID  770554 non-null  float64       
 7   Country      770554 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 62.8+ MB


In [18]:
# Confirm Quantity and Price are both strictly positive
df_clean['Quantity'].min(), df_clean['Price'].min()

(np.int64(1), np.float64(0.03))

In [19]:
# Verify the confirmed phantom orders are gone (string comparison: StockCode mixes ints and strings)
inv = df_clean['Invoice'].astype(str)
code = df_clean['StockCode'].astype(str)

print("12346 storage jar order (541431), rows remaining:",
      ((inv == '541431') & (code == '23166')).sum(), "(expected 0)")
print("16446 Little Birdie order (581483), rows remaining:",
      ((inv == '581483') & (code == '23843')).sum(), "(expected 0)")

# 15098: three same-day orders of 22502 (60 @ £4.95, 60 @ £649.50, 1 @ £649.50), then 60 cancelled.
# LIFO offsets the two most recent orders, so exactly 1 unit should remain on the £649.50 invoices.
basket = (df_clean['Customer ID'] == 15098) & (code == '22502') & inv.isin(['556444', '556446'])
print("15098 units remaining on the £649.50 invoices:",
      df_clean.loc[basket, 'Quantity'].sum(), "(expected 1)")

12346 storage jar order (541431), rows remaining: 0 (expected 0)
16446 Little Birdie order (581483), rows remaining: 0 (expected 0)
15098 units remaining on the £649.50 invoices: 1 (expected 1)


In [20]:
cols = ['Invoice', 'Quantity', 'Price', 'InvoiceDate']
mask_raw = (df['Customer ID'] == 15098) & (df['StockCode'].astype(str) == '22502')
mask_clean = (df_clean['Customer ID'] == 15098) & (df_clean['StockCode'].astype(str) == '22502')

print("Before netting:")
print(df[mask_raw][cols])
print("\nAfter netting:")
print(df_clean[mask_clean][cols])

Before netting:
        Invoice  Quantity   Price         InvoiceDate
717209   556442        60    4.95 2011-06-10 15:22:00
717210   556444        60  649.50 2011-06-10 15:28:00
717221   556446         1  649.50 2011-06-10 15:33:00
717231  C556448       -60    4.95 2011-06-10 15:39:00

After netting:
       Invoice  Quantity   Price         InvoiceDate
717209  556442        60    4.95 2011-06-10 15:22:00
717210  556444         1  649.50 2011-06-10 15:28:00


Customer 15098 placed three orders for product 22502 within 11 minutes on 2011-06-10 (60 units at £4.95, then 60 units at £649.50, then 1 unit at £649.50), followed by a cancellation of 60 units keyed at £4.95. LIFO offsets the cancellation against the two most recent orders, leaving 60 units at £4.95 and 1 unit at £649.50 (£946.50 in total).

Matching on price would instead have offset the £4.95 order exactly and kept £39,619.50 of the £649.50 orders, including an implausible 60-unit order worth £38,970. This is the clearest evidence in the data for using recency rather than price as the matching signal.

Netting removes only what cancellation records show was reversed, so the remaining unit at £649.50 stays in the data. Whether it was a genuine purchase or an uncancelled pricing error cannot be determined from the records; this is noted as a limitation.

In [21]:
# Final audit: any StockCode outside the standard product format that survived cleaning
codes = df_clean['StockCode'].astype(str)
remaining = df_clean[~codes.str.fullmatch(r'\d{5}[A-Za-z]{0,2}')]

print("Non-standard codes remaining:", sorted(remaining['StockCode'].astype(str).unique()))
print("Expected: ['SP1002'] (a genuine product in a non-standard code format)")
print("Minimum Price after cleaning:", df_clean['Price'].min())

Non-standard codes remaining: ['SP1002']
Expected: ['SP1002'] (a genuine product in a non-standard code format)
Minimum Price after cleaning: 0.03


### Summary

Starting from 1,067,371 raw transaction records, the following cleaning steps were applied:

| Step | Rows Removed | Rows Remaining |
|---|---|---|
| Remove exact duplicate rows | 34,335 | 1,033,036 |
| Drop missing `Customer ID` | 235,151 | 797,885 |
| Drop non-product `StockCode` entries | 3,644 | 794,241 |
| Drop `Price <= 0` | 60 | 794,181 |
| Net cancellations against purchases (LIFO) | 23,627 | 770,554 |

**Total rows removed:** 296,817 (27.8% of the original dataset)

**Negative quantities without a cancellation prefix.** The 3,393 negative-quantity rows without a 'C' invoice all lack a `Customer ID` and have a price of £0. Their descriptions record stock adjustments ("damaged", "smashed", "missing", "check"), not customer returns. They are removed by the missing `Customer ID` filter and never reach customer-level analysis.

**Non-product codes.** Codes outside the standard product format (five digits, optionally followed by one or two letters) were reviewed by description. Postage, carriage, discounts, bank charges, manual entries, manual accounting adjustments (`ADJUST`, `ADJUST2`), test products (`TEST001`, `TEST002`) and a £0.001 placeholder (`PADS`) are excluded. `SP1002`, a genuine product with a non-standard code, is kept. The list is defined once in `src/config.py`.

**Cancellation netting.** A cancellation reverses an earlier purchase, so dropping only the cancellation line would keep revenue that was never received. Each cancellation is instead offset against the same customer's earlier purchases of the same product, most recent first (LIFO). Of 17,602 cancellation lines, 15,877 were fully offset, 126 partly offset, and 1,599 had no earlier purchase in the data to offset (£42,066 of cancelled value, discarded). Netting removed £730,406 of recorded purchase value, about 4.2% of gross product sales: 6,025 purchase rows were removed entirely and a further 9,683 had their quantity reduced.

The largest effects come from a few orders cancelled within minutes of being placed: customer 12346's £77,183.60 storage-jar order and customer 16446's £168,469.60 "Paper Craft, Little Birdie" order. The second falls in the modelling holdout window, where it would otherwise have become that customer's entire 6-month target value.

**Why recency, not price, is the matching signal.** Customer 15098 placed three same-day orders for one product (60 units at £4.95, 60 at £649.50, 1 at £649.50), then cancelled 60 units keyed at £4.95. Matching on price would have offset the £4.95 order and kept £39,619.50 of the £649.50 orders. LIFO offsets the two most recent orders and leaves £946.50, a plausible same-day wholesale purchase. Observed price mismatches elsewhere were either symmetric noise or similar pricing corrections.

**Limitation.** Netting removes only what cancellation records show was reversed. Pricing or entry errors that were never cancelled, such as the single remaining unit at £649.50 for customer 15098, stay in the data.

**Comparison with the earlier cleaning rule.** The previous version of this notebook dropped rows with `Quantity <= 0` (removing cancellation lines but keeping the purchases they reversed) and used a shorter non-product list, producing 776,641 rows. The final dataset has 6,087 fewer: 6,025 purchases netted away by cancellations and 62 purchase lines of the newly excluded non-product codes. A further 9,683 purchases have reduced quantities; every other row is identical.

**Verification.** The confirmed cancelled orders for customers 12346 and 16446 are fully removed, exactly one unit remains on customer 15098's £649.50 invoices, the only non-standard code remaining is `SP1002`, and the minimum price is £0.03.

The cleaned dataset (`df_clean`) contains 770,554 transaction records across 8 columns, with no missing values, strictly positive `Quantity` and `Price`, no non-product entries, and no cancelled revenue. It is saved to disk as a checkpoint for downstream feature engineering and modelling.

In [22]:
# Save the cleaned dataset as a checkpoint for downstream notebooks
df_clean.to_csv("../data/online_retail_clean.csv", index=False)